pkg/plugin/search_api.go

```diff
@@ -0,0 +1,174 @@
+package plugin
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/criblcloud/search-datasource/pkg/models"
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+)
+
+func NewSearchAPI(settings *models.PluginSettings) *SearchAPI {
+	return &SearchAPI{
+		Settings:    settings,
+		BearerToken: nil,
+	}
+}
+
+type SearchAPI struct {
+	Settings    *models.PluginSettings
+	BearerToken *BearerToken
+}
+
+type SearchQueryResult struct {
+	Header map[string]interface{}
+	Events []map[string]interface{}
+}
+
+// Run a search query and return the header event + result events.  The queryParams arg is
+// expected to have params such as query + earlieset + latest, or a savedSearchId, and
+// any offset + limit as needed.  This simply makes the API request and parses the response.
+func (api *SearchAPI) RunQueryAndGetResults(queryParams *url.Values) (*SearchQueryResult, error) {
+	responseBytes, err := api.doGET("/api/v1/m/default_search/search/query", queryParams)
+	if err != nil {
+		return nil, err
+	}
+	// The response is NDJSON, one header "event" plus result events
+	lines := strings.Split(string(responseBytes), "\n")
+	result := SearchQueryResult{}
+	for idx, line := range lines {
+		if len(line) == 0 {
+			continue
+		}
+		var event map[string]interface{}
+		if err := json.Unmarshal([]byte(line), &event); err != nil {
+			return nil, fmt.Errorf("failed to parse json at line %d: %s", idx+1, line)
+		}
+		if idx == 0 {
+			result.Header = event
+		} else {
+			result.Events = append(result.Events, event)
+		}
+	}
+	return &result, nil
+}
+
+// Load the list of saved search IDs available to the user corresponding to the API creds.
+// This can be used to populate a dropdown to make it easy for the user to pick one.
+// Returns a list of saved search IDs.
+func (api *SearchAPI) LoadSavedSearchIds() ([]string, error) {
+	responseBytes, err := api.doGET("/api/v1/m/default_search/search/saved", nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load saved search ids: %v", err.Error())
+	}
+	var data struct {
+		Items []map[string]any `json:"items"`
+	}
+	err = json.Unmarshal(responseBytes, &data)
+	if err != nil {
```

Comment:
nit: I guess you could collapse this into `if err := json.Unmarshal(responseBytes, &data); err != nil`

Reply:
Definitely, will do!

Reply:
@itsalvinho in this case you can't "initialize" `err` directly with `:=` since has already been defined.  But you CAN collapse into one line (which I'll do).